webhookv3/webhook_event.go

```diff
@@ -0,0 +1,74 @@
+package webhookv3
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/PagerDuty/go-pagerduty"
+)
+
+type OutboundEventData struct {
+	Object map[string]interface{}
+	// The raw json data for use in structured unmarshalling.
+	RawData json.RawMessage
+}
+
+type OutboundEvent struct {
+	ID           string                  `json:"id"`
+	EventType    string                  `json:"event_type"`
+	ResourceType string                  `json:"resource_type"`
+	OccurredAt   string                  `json:"occurred_at"`
+	Agent        *pagerduty.APIReference `json:"agent"`
+	Data         *OutboundEventData      `json:"data"`
+}
+
+type WebhookPayload struct {
+	Event OutboundEvent `json:"event"`
+}
+
+func (e *OutboundEventData) UnmarshalJSON(data []byte) error {
+	if err := json.Unmarshal(data, &e.Object); err != nil {
+		return err
+	}
+
+	if err := e.RawData.UnmarshalJSON(data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (oe *OutboundEvent) GetEventDataValue(keys ...string) (string, error) {
```

Comment:
Ah so I guess I should have provided a bit more color than I did in the PR description. A few items of background: 

* V3 Webhooks are still in the early stages of life at PagerDuty and we are routinely adding new [event types](https://developer.pagerduty.com/docs/ZG9jOjExMDI5NTkw-v3-overview#event-types) / [event data types](https://developer.pagerduty.com/docs/ZG9jOjExMDI5NTkw-v3-overview#event-data-types).
* The [event data types](https://developer.pagerduty.com/docs/ZG9jOjExMDI5NTkw-v3-overview#event-data-types) delivered in Outbound Events are not exactly equal to their corresponding REST API representation. We are trying our best for them to be a subset of these resources.
* Historically, PagerDuty has not always provided concrete static types in all of our client libraries.

This method is not meant to be mutually exclusive with a set of methods that describe concrete types as described in https://github.com/PagerDuty/go-pagerduty/issues/367#issuecomment-945282862. [I have actually started a bit of work on that as well.] Instead, this method is meant to allow consumers to work with V3 Webhooks in a generic way, particularly in situations where the full structured types may not have been implemented quite yet.

While there are examples of how to use this function in `examples/webhooks/webhook_server.go` as well as in the tests, this whole file was missing some `godoc` and that has been added. Have a look and let me know what you think...

Reply:
Okay, that makes it a bit more clear to me. Although I do have one question:

> Historically, PagerDuty has not always provided concrete static types in all of our client libraries.

How many of your client libraries are for statically typed languages? If they've been more focused towards the Perl, Ruby, Python, and JS communities, it doesn't surprise me that there haven't been static types defined.

With the cost of doing a major version bump in Go, especially because we're already at v1, we should be extra thoughtful before increasing the API surface area because we're probably going to be stuck with it for quite awhile. I say this because I don't know that I'm aligned with expanding the API surface area in this way, unrelated to maintenance cost of adding this to the public API.

Instead of expanding our API in this way, in cases where we've yet to define a `struct` type, I'd instead advocate for consumers to define their own struct (bonus points if they contribute it back to us). The reason I'm leaning towards this path, is that whether they want to define their own struct, or use `GetEventDataValue`, they will need to inspect the API documentation to parse the data.

Further compounding this, `GetEventDataValue` only returns `string` values, but there are webhook fields that are non-string values. So not only will they need to know which field(s) to extract from the payload, but they will also need to know the value's type and handle any conversions themselves.

Finally, if they want to use multiple values from within the webhook, they will need to extract those values individually, possibly convert some from a `string` to a more appropriate type (`int`, `bool`, etc.), and then put them inside of some container that allows them to be passed around. I suspect most folks would define a `struct` type for this container, if one wasn't already present in `package webhookv3`.

So considering all of these factors that would drive a consumer to define a `struct` type anyway, I am not personally in favor of this API being added to the `webhookv3` package. Instead, I'd like to see us implement the most common types, commit to adding new types and fields as they are documented, and finally in cases where we have not defined the type yet encourage consumers to define their own struct types as needed and contribute them back as it makes sense.

Reply:
While I am not going to attempt to unpack _all_ of this here, I would be happy to Zoom or exchange a few emails if you would like to discuss further.

We would really like to keep this method in place and would appreciate any feedback you might have around how to make it better.

That being said, see what you think about https://github.com/ChezCrawford/go-pagerduty/pull/2 for also adding support for structured events into the mix at the same time.